File: src/main/java/org/oewntk/model/KeyF.java
/*
 * Copyright (c) 2021. Bernard Bou.
 */

package org.oewntk.model;

import java.util.function.Function;

public interface KeyF<R> extends Function<CoreModel, R>
{
	interface MonoValued extends KeyF<Lex>
	{
	}

	interface MultiValued extends KeyF<Lex[]>
	{
	}

	class W_P_functional<L extends Function<Lex, String>, P extends Function<Lex, Character>> extends Key.W_P
	{
		final L wordExtractor;

		final P posTypeExtractor;

		private W_P_functional(final String word, final char posType, final L wordExtractor, final P posTypeExtractor)
		{
			super(word, posType);
			this.wordExtractor = wordExtractor;
			this.posTypeExtractor = posTypeExtractor;
		}

		public static <L extends Function<Lex, String>, P extends Function<Lex, Character>> W_P_functional<L, P> of(final Lex lex, final L wordExtractor, final P posTypeExtractor)
		{
			return new W_P_functional<>(wordExtractor.apply(lex), posTypeExtractor.apply(lex), wordExtractor, posTypeExtractor);
		}

		public static class Mono<L extends Function<Lex, String>, P extends Function<Lex, Character>> extends W_P_functional<L, P> implements MonoValued
		{
			public static <L extends Function<Lex, String>, P extends Function<Lex, Character>> Mono of(final L wordExtractor, final P posTypeExtractor, final Lex lex)
			{
				return new Mono<>(wordExtractor.apply(lex), posTypeExtractor.apply(lex), wordExtractor, posTypeExtractor);
			}

			public static <L extends Function<Lex, String>, P extends Function<Lex, Character>> Mono<L, P> from(final L wordExtractor, final P posTypeExtractor, final String word, final char posType)
			{
				return new Mono<>(word, posType, wordExtractor, posTypeExtractor);
			}

			private Mono(final String word, final char posType, final L wordExtractor, final P posTypeExtractor)
			{
				super(word, posType, wordExtractor, posTypeExtractor);
			}

			@Override
			public Lex apply(final CoreModel model)
			{
				return Finder.getLexesHaving(model, word, posType, wordExtractor, posTypeExtractor).findFirst().orElseThrow(IllegalArgumentException::new);
			}
		}

		public static class Multi<L extends Function<Lex, String>, P extends Function<Lex, Character>> extends W_P_functional<L, P> implements MultiValued
		{
			public static <L extends Function<Lex, String>, P extends Function<Lex, Character>> Multi of(final L wordExtractor, final P posTypeExtractor, final Lex lex)
			{
				return new Multi<>(wordExtractor.apply(lex), posTypeExtractor.apply(lex), wordExtractor, posTypeExtractor);
			}

			public static <L extends Function<Lex, String>, P extends Function<Lex, Character>> Multi<L, P> from(final L wordExtractor, final P posTypeExtractor, final String word, final char posType)
			{
				return new Multi<>(word, posType, wordExtractor, posTypeExtractor);
			}

			private Multi(final String word, final char posType, final L wordExtractor, final P posTypeExtractor)
			{
				super(word, posType, wordExtractor, posTypeExtractor);
			}

			@Override
			public Lex[] apply(final CoreModel model)
			{
				return Finder.getLexesHaving(model, word, posType, wordExtractor, posTypeExtractor).toArray(Lex[]::new);
			}
		}
	}

	class W_P_A_functional<L extends Function<Lex, String>, P extends Function<Lex, Character>> extends Key.W_P_A
	{
		final L wordExtractor;

		final P posTypeExtractor;

		private W_P_A_functional(final String word, final char posType, final Pronunciation[] pronunciations, final L wordExtractor, final P posTypeExtractor)
		{
			super(word, posType, pronunciations);
			this.wordExtractor = wordExtractor;
			this.posTypeExtractor = posTypeExtractor;
		}

		public static <L extends Function<Lex, String>, P extends Function<Lex, Character>> W_P_A_functional<L, P> of(final Lex lex, final L wordExtractor, final P posTypeExtractor)
		{
			return new W_P_A_functional<>(wordExtractor.apply(lex), posTypeExtractor.apply(lex), lex.getPronunciations(), wordExtractor, posTypeExtractor);
		}

		public static class Mono<L extends Function<Lex, String>, P extends Function<Lex, Character>> extends W_P_A_functional<L, P> implements MonoValued
		{
			public static <L extends Function<Lex, String>, P extends Function<Lex, Character>> Mono<L, P> of(final L wordExtractor, final P posTypeExtractor, final Lex lex)
			{
				return new Mono<>(wordExtractor.apply(lex), posTypeExtractor.apply(lex), lex.getPronunciations(), wordExtractor, posTypeExtractor);
			}

			public static <L extends Function<Lex, String>, P extends Function<Lex, Character>> Mono<L, P> from(final L wordExtractor, final P posTypeExtractor, final String word, final char posType, final Pronunciation... pronunciations)
			{
				return new Mono<>(word, posType, pronunciations, wordExtractor, posTypeExtractor);
			}

			private Mono(final String word, final char posType, final Pronunciation[] pronunciations, final L wordExtractor, final P posTypeExtractor)
			{
				super(word, posType, pronunciations, wordExtractor, posTypeExtractor);
			}

			@Override
			public Lex apply(final CoreModel model)
			{
				return Finder.getLexesHavingPronunciations(Finder.getLexesHaving(model, word, posType, wordExtractor, posTypeExtractor), pronunciations).findFirst().orElseThrow(IllegalArgumentException::new);
			}
		}

		public static class Multi<L extends Function<Lex, String>, P extends Function<Lex, Character>> extends W_P_A_functional<L, P> implements MultiValued
		{
			public static <L extends Function<Lex, String>, P extends Function<Lex, Character>> Multi<L, P> of(final L wordExtractor, final P posTypeExtractor, final Lex lex)
			{
				return new Multi<>(wordExtractor.apply(lex), posTypeExtractor.apply(lex), lex.getPronunciations(), wordExtractor, posTypeExtractor);
			}

			public static <L extends Function<Lex, String>, P extends Function<Lex, Character>> Multi<L, P> from(final L wordExtractor, final P posTypeExtractor, final String word, final char posType, final Pronunciation... pronunciations)
			{
				return new Multi<>(word, posType, pronunciations, wordExtractor, posTypeExtractor);
			}

			private Multi(final String word, final char posType, final Pronunciation[] pronunciations, final L wordExtractor, final P posTypeExtractor)
			{
				super(word, posType, pronunciations, wordExtractor, posTypeExtractor);
			}

			@Override
			public Lex[] apply(final CoreModel model)
			{
				return Finder.getLexesHavingPronunciations(Finder.getLexesHaving(model, word, posType, wordExtractor, posTypeExtractor), pronunciations).toArray(Lex[]::new);
			}
		}
	}

	class W_P_D_functional<L extends Function<Lex, String>, P extends Function<Lex, Character>> extends Key.W_P_D
	{
		final L wordExtractor;

		final P posTypeExtractor;

		private W_P_D_functional(final String word, final char posType, final String discriminant, final L wordExtractor, final P posTypeExtractor)
		{
			super(word, posType, discriminant);
			this.wordExtractor = wordExtractor;
			this.posTypeExtractor = posTypeExtractor;
		}

		public static <L extends Function<Lex, String>, P extends Function<Lex, Character>> W_P_D_functional<L, P> of(final Lex lex, final L wordExtractor, final P posTypeExtractor)
		{
			return new W_P_D_functional<>(wordExtractor.apply(lex), posTypeExtractor.apply(lex), lex.getDiscriminant(), wordExtractor, posTypeExtractor);
		}

		public static class Mono<L extends Function<Lex, String>, P extends Function<Lex, Character>> extends W_P_D_functional<L, P> implements MonoValued
		{
			public static <L extends Function<Lex, String>, P extends Function<Lex, Character>> Mono<L, P> from(final L wordExtractor, final P posTypeExtractor, final String word, final char posType, final String discriminant)
			{
				return new Mono<>(word, posType, discriminant, wordExtractor, posTypeExtractor);
			}

			private Mono(final String word, final char posType, final String discriminant, final L wordExtractor, final P posTypeExtractor)
			{
				super(word, posType, discriminant, wordExtractor, posTypeExtractor);
			}

			@Override
			public Lex apply(final CoreModel model)
			{
				return Finder.getLexesHavingDiscriminant(Finder.getLexesHaving(model, word, posType, wordExtractor, posTypeExtractor), discriminant).findFirst().orElseThrow(IllegalArgumentException::new);
			}
		}

		public static class Multi<L extends Function<Lex, String>, P extends Function<Lex, Character>> extends W_P_D_functional<L, P> implements MultiValued
		{
			public static <L extends Function<Lex, String>, P extends Function<Lex, Character>> Multi<L, P> from(final L wordExtractor, final P posTypeExtractor, final String word, final char posType, final String discriminant)
			{
				return new Multi<>(word, posType, discriminant, wordExtractor, posTypeExtractor);
			}

			private Multi(final String word, final char posType, final String discriminant, final L wordExtractor, final P posTypeExtractor)
			{
				super(word, posType, discriminant, wordExtractor, posTypeExtractor);
			}


			@Override
			public Lex[] apply(final CoreModel model)
			{
				return Finder.getLexesHavingDiscriminant(Finder.getLexesHaving(model, word, posType, wordExtractor, posTypeExtractor), discriminant).toArray(Lex[]::new);
			}
		}
	}
}

File: src/main/java/org/oewntk/model/KeyF.kt
/*
 * Copyright (c) 2021-2024. Bernard Bou.
 */
package org.oewntk.model

import org.oewntk.model.Key.*

/**
 * Keys extended with a functional interface that extract data R from model
 *
 * @param R result
 */
interface KeyF<R> : Key, (CoreModel) -> R {

    fun toLongString(): String

    /**
     * Interface for single value output
     */
    interface MonoValued : KeyF<Lex>

    /**
     * Interface for multiple values (array) output
     */
    interface MultiValued : KeyF<Array<Lex>>

    /**
     * (Lemma, Category)
     *
     * @param lemma                 lemma or LC lemma
     * @param category              category: part-of-speech or type
     * @property lemmaExtractor     lemma extractor function from lex
     * @property categoryExtractor  category extractor function from lex
     */
    open class FuncFromLemmaCategory private constructor(
        override var lemma: Lemma,
        override var category: Category,
        val lemmaExtractor: (Lex) -> Lemma,
        val categoryExtractor: (Lex) -> Category,

        ) : FromLemmaCategory(lemma, category) {

        override fun toLongString(): String = "KEYF ${javaClass.simpleName} LC_${lemmaExtractor.toCaseSensitiveOrLowerCased()}_${categoryExtractor.toTypeOrPos()} $this"

        /**
         * Functional part that yields single value
         */
        class Mono private constructor(
            lemma: Lemma,
            category: Category,
            lemmaExtractor: (Lex) -> Lemma,
            categoryExtractor: (Lex) -> Category,
        ) : FuncFromLemmaCategory(lemma, category, lemmaExtractor, categoryExtractor), MonoValued {

            override fun invoke(model: CoreModel): Lex {
                return Finder.getLexesHaving(model, lemma, category, lemmaExtractor, categoryExtractor)
                    .firstOrNull() ?: throw IllegalArgumentException()
            }

            companion object {

                fun of(
                    lemmaExtractor: (Lex) -> Lemma,
                    categoryExtractor: (Lex) -> Category,
                    lex: Lex,
                ): Mono {
                    return Mono(
                        lemmaExtractor.invoke(lex),
                        categoryExtractor.invoke(lex),
                        lemmaExtractor,
                        categoryExtractor
                    )
                }

                fun from(
                    lemmaExtractor: (Lex) -> Lemma,
                    categoryExtractor: (Lex) -> Category,
                    lemma: Lemma,
                    category: Category,
                ): Mono {
                    return Mono(lemma, category, lemmaExtractor, categoryExtractor)
                }
            }
        }

        /**
         * Functional part that yields multiple values (array)
         */
        class Multi private constructor(
            lemma: Lemma,
            category: Category,
            lemmaExtractor: (Lex) -> Lemma,
            categoryExtractor: (Lex) -> Category,
        ) : FuncFromLemmaCategory(lemma, category, lemmaExtractor, categoryExtractor), MultiValued {

            override fun invoke(model: CoreModel): Array<Lex> {
                return Finder.getLexesHaving(model, lemma, category, lemmaExtractor, categoryExtractor)
                    .toList()
                    .toTypedArray()
            }

            companion object {

                fun of(
                    lemmaExtractor: (Lex) -> Lemma,
                    categoryExtractor: (Lex) -> Category,
                    lex: Lex,
                ): Multi {
                    return Multi(
                        lemmaExtractor.invoke(lex),
                        categoryExtractor.invoke(lex),
                        lemmaExtractor,
                        categoryExtractor
                    )
                }

                fun from(
                    lemmaExtractor: (Lex) -> Lemma,
                    categoryExtractor: (Lex) -> Category,
                    lemma: Lemma,
                    category: Category,
                ): Multi {
                    return Multi(lemma, category, lemmaExtractor, categoryExtractor)
                }
            }
        }

        companion object {

            fun of(
                lex: Lex,
                lemmaExtractor: (Lex) -> Lemma,
                categoryExtractor: (Lex) -> Category,
            ): FuncFromLemmaCategory {
                return FuncFromLemmaCategory(
                    lemmaExtractor.invoke(lex),
                    categoryExtractor.invoke(lex),
                    lemmaExtractor,
                    categoryExtractor
                )
            }
        }
    }

    /**
     * (Lemma, Category, Pronunciations)
     *
     * @param lemma                lemma or LC lemma
     * @param category             category: part-of-speech or type
     * @param pronunciations       pronunciations
     * @property lemmaExtractor    lemma extractor function from lex
     * @property categoryExtractor category extractor function from lex
     */
    open class FuncUsingPronunciation private constructor(
        lemma: Lemma,
        category: Category,
        pronunciations: Set<Pronunciation>?,
        val lemmaExtractor: (Lex) -> Lemma,
        val categoryExtractor: (Lex) -> Category,
    ) : UsingPronunciation(lemma, category, pronunciations) {

        override fun toLongString(): String = "KEYF ${javaClass.simpleName} LCP_${lemmaExtractor.toCaseSensitiveOrLowerCased()}_${categoryExtractor.toTypeOrPos()} $this"

        /**
         * Functional part that yields single value
         */
        class Mono private constructor(
            lemma: Lemma,
            category: Category,
            pronunciations: Set<Pronunciation>?,
            lemmaExtractor: (Lex) -> Lemma,
            categoryExtractor: (Lex) -> Category,
        ) : FuncUsingPronunciation(lemma, category, pronunciations, lemmaExtractor, categoryExtractor), MonoValued {

            override fun invoke(model: CoreModel): Lex {
                return Finder.getLexesHavingPronunciations(
                    Finder.getLexesHaving(
                        model,
                        lemma,
                        category,
                        lemmaExtractor,
                        categoryExtractor
                    ), pronunciations
                ).firstOrNull() ?: throw IllegalArgumentException()
            }

            companion object {

                fun of(
                    lemmaExtractor: (Lex) -> Lemma,
                    categoryExtractor: (Lex) -> Category,
                    lex: Lex,
                ): Mono {
                    return Mono(
                        lemmaExtractor.invoke(lex),
                        categoryExtractor.invoke(lex),
                        lex.pronunciations,
                        lemmaExtractor,
                        categoryExtractor
                    )
                }

                fun from(
                    lemmaExtractor: (Lex) -> Lemma,
                    categoryExtractor: (Lex) -> Category,
                    lemma: Lemma,
                    category: Category,
                    pronunciations: Set<Pronunciation>,
                ): Mono {
                    return Mono(lemma, category, pronunciations, lemmaExtractor, categoryExtractor)
                }
            }
        }

        /**
         * Functional part that yields multiple values (array)
         */
        class Multi private constructor(
            lemma: Lemma,
            category: Category,
            pronunciations: Set<Pronunciation>?,
            lemmaExtractor: (Lex) -> Lemma,
            categoryExtractor: (Lex) -> Category,
        ) : FuncUsingPronunciation(lemma, category, pronunciations, lemmaExtractor, categoryExtractor), MultiValued {

            override fun invoke(model: CoreModel): Array<Lex> {
                return Finder.getLexesHavingPronunciations(
                    Finder.getLexesHaving(
                        model,
                        lemma,
                        category,
                        lemmaExtractor,
                        categoryExtractor
                    ), pronunciations
                )
                    .toList()
                    .toTypedArray()
            }

            companion object {

                fun of(
                    lemmaExtractor: (Lex) -> Lemma,
                    categoryExtractor: (Lex) -> Category,
                    lex: Lex,
                ): Multi {
                    return Multi(
                        lemmaExtractor.invoke(lex),
                        categoryExtractor.invoke(lex),
                        lex.pronunciations,
                        lemmaExtractor,
                        categoryExtractor
                    )
                }

                fun from(
                    lemmaExtractor: (Lex) -> Lemma,
                    categoryExtractor: (Lex) -> Category,
                    lemma: Lemma,
                    category: Category,
                    pronunciations: Set<Pronunciation>,
                ): Multi {
                    return Multi(lemma, category, pronunciations, lemmaExtractor, categoryExtractor)
                }
            }
        }

        companion object {

            fun of(
                lex: Lex,
                lemmaExtractor: (Lex) -> Lemma,
                categoryExtractor: (Lex) -> Category,
            ): FuncUsingPronunciation {
                return FuncUsingPronunciation(
                    lemmaExtractor.invoke(lex),
                    categoryExtractor.invoke(lex),
                    lex.pronunciations,
                    lemmaExtractor,
                    categoryExtractor
                )
            }
        }
    }

    /**
     * (Lemma, Category, Discriminant) - Shallow key
     *
     * @param lemma             lemma or LC lemma
     * @param category          category: part-of-speech or type
     * @param discriminant      discriminant
     * @property lemmaExtractor lemma extractor function from lex
     * @property categoryExtractor category extractor function from lex
     */
    open class FuncUsingDiscriminant private constructor(
        lemma: Lemma,
        category: Category,
        discriminant: Discriminant?,
        val lemmaExtractor: (Lex) -> Lemma,
        val categoryExtractor: (Lex) -> Category,
    ) : UsingDiscriminant(lemma, category, discriminant) {

        override fun toLongString(): String = "KEYF ${javaClass.simpleName} LCD_${lemmaExtractor.toCaseSensitiveOrLowerCased()}_${categoryExtractor.toTypeOrPos()} $this"

        /**
         * Functional part that yields single value
         */
        @Suppress("unused")
        class Mono private constructor(
            lemma: Lemma,
            category: Category,
            discriminant: Discriminant?,
            lemmaExtractor: (Lex) -> Lemma,
            categoryExtractor: (Lex) -> Category,
        ) : FuncUsingDiscriminant(lemma, category, discriminant, lemmaExtractor, categoryExtractor), MonoValued {

            override fun invoke(model: CoreModel): Lex {
                return Finder.getLexesHavingDiscriminant(
                    Finder.getLexesHaving(
                        model,
                        lemma,
                        category,
                        lemmaExtractor,
                        categoryExtractor
                    ), discriminant
                )
                    .firstOrNull() ?: throw IllegalArgumentException()
            }

            companion object {

                fun from(
                    lemmaExtractor: (Lex) -> Lemma,
                    categoryExtractor: (Lex) -> Category,
                    lemma: Lemma,
                    category: Category,
                    discriminant: Discriminant?,
                ): Mono {
                    return Mono(lemma, category, discriminant, lemmaExtractor, categoryExtractor)
                }
            }
        }

        /**
         * Functional part that yields multiple values (array)
         */
        class Multi private constructor(
            lemma: Lemma,
            category: Category,
            discriminant: Discriminant?,
            lemmaExtractor: (Lex) -> Lemma,
            categoryExtractor: (Lex) -> Category,
        ) : FuncUsingDiscriminant(lemma, category, discriminant, lemmaExtractor, categoryExtractor), MultiValued {

            override fun invoke(model: CoreModel): Array<Lex> {
                return Finder.getLexesHavingDiscriminant(
                    Finder.getLexesHaving(
                        model,
                        lemma,
                        category,
                        lemmaExtractor,
                        categoryExtractor
                    ), discriminant
                )
                    .toList()
                    .toTypedArray()
            }

            companion object {

                fun from(
                    lemmaExtractor: (Lex) -> Lemma,
                    categoryExtractor: (Lex) -> Category,
                    lemma: Lemma,
                    category: Category,
                    discriminant: Discriminant?,
                ): Multi {
                    return Multi(lemma, category, discriminant, lemmaExtractor, categoryExtractor)
                }
            }
        }

        companion object {

            fun of(
                lex: Lex,
                lemmaExtractor: (Lex) -> Lemma,
                categoryExtractor: (Lex) -> Category,
            ): FuncUsingDiscriminant {
                return FuncUsingDiscriminant(
                    lemmaExtractor.invoke(lex),
                    categoryExtractor.invoke(lex),
                    lex.discriminant,
                    lemmaExtractor,
                    categoryExtractor
                )
            }
        }
    }

    companion object {

        private const val DUMMY_UPPER = "CASE"

        private val dummyLex = Lex(DUMMY_UPPER, Category.S.toString())

        /**
         * Name a lemma extractor (by applying dummy data)
         *
         * @receiver lemma extractor
         * @return name
         */
        fun ((Lex) -> Lemma).toCaseSensitiveOrLowerCased(): String {
            return if (this.invoke(dummyLex) == DUMMY_UPPER) "cs" else "lc"
        }

        /**
         * Name a pos/type category extractor (by applying dummy data)
         *
         * @receiver category extractor
         * @return name
         */
        fun ((Lex) -> Category).toTypeOrPos(): String {
            return if (this.invoke(dummyLex) == Category.S) "t" else "p"
        }
    }
}
